util/topsql/reporter/datamodel.go

```diff
@@ -0,0 +1,685 @@
+	return &tipb.TopSQLRecordItem{
+		TimestampSec:    i.timestamp,
+		CpuTimeMs:       i.cpuTimeMs,
+		StmtExecCount:   i.stmtStats.ExecCount,
+		StmtKvExecCount: i.stmtStats.KvStatsItem.KvExecCount,
+		// TODO: add duration
+		// StmtDurationSumNs: i.stmtStats.DurationSumNs,
+		// Convert more indicators here.
+	}
+}
+
+var _ sort.Interface = &tsItems{}
+
+// tsItems is a sortable list of tsItem, sort by tsItem.timestamp (asc).
+type tsItems []tsItem
+
+func (ts tsItems) Len() int {
+	return len(ts)
+}
+
+func (ts tsItems) Less(i, j int) bool {
+	return ts[i].timestamp < ts[j].timestamp
+}
+
+func (ts tsItems) Swap(i, j int) {
+	ts[i], ts[j] = ts[j], ts[i]
+}
+
+func (ts tsItems) sorted() bool {
+	for n := 0; n < len(ts)-1; n++ {
+		if ts[n].timestamp > ts[n+1].timestamp {
+			return false
+		}
+	}
+	return true
+}
+
+// toProto converts the tsItems to the corresponding protobuf representation.
+func (ts tsItems) toProto() []*tipb.TopSQLRecordItem {
+	capacity := len(ts)
+	if capacity == 0 {
+		return nil
+	}
+	items := make([]*tipb.TopSQLRecordItem, 0, capacity)
+	for _, i := range ts {
+		items = append(items, i.toProto())
+	}
+	return items
+}
+
+var _ sort.Interface = &record{}
+
+// record represents the cumulative tsItem in current minute window.
+// record do not guarantee the tsItems is sorted by timestamp when there is a time jump backward.
+// record is also sortable, and the tsIndex will be updated while sorting the internal tsItems.
+type record struct {
+	sqlDigest      []byte
+	planDigest     []byte
+	totalCPUTimeMs uint64
+	tsItems        tsItems
+
+	// tsIndex is used to quickly find the corresponding tsItems index through timestamp.
+	tsIndex map[uint64]int // timestamp => index of tsItems
+}
+
+func newRecord(sqlDigest, planDigest []byte) *record {
+	listCap := variable.TopSQLVariable.ReportIntervalSeconds.Load()/variable.TopSQLVariable.PrecisionSeconds.Load() + 1
+	return &record{
+		sqlDigest:  sqlDigest,
+		planDigest: planDigest,
+		tsItems:    make(tsItems, 0, listCap),
+		tsIndex:    make(map[uint64]int, listCap),
+	}
+}
+
+func (r *record) Len() int {
+	return r.tsItems.Len()
+}
+
+func (r *record) Less(i, j int) bool {
+	return r.tsItems.Less(i, j)
+}
+
+func (r *record) Swap(i, j int) {
+	// before swap:
+	//     timestamps: [10000, 10001, 10002]
+	//        tsIndex: [10000 => 0, 10001 => 1, 10002 => 2]
+	//
+	// let i = 0, j = 1
+	// after swap tsIndex:
+	//     timestamps: [10000, 10001, 10002]
+	//        tsIndex: [10000 => 1, 10001 => 0, 10002 => 2]
+	//
+	// after swap tsItems:
+	//     timestamps: [10001, 10000, 10002]
+	//        tsIndex: [10000 => 1, 10001 => 0, 10002 => 2]
+	r.tsIndex[r.tsItems[i].timestamp], r.tsIndex[r.tsItems[j].timestamp] = r.tsIndex[r.tsItems[j].timestamp], r.tsIndex[r.tsItems[i].timestamp]
+	r.tsItems.Swap(i, j)
+}
+
+// appendCPUTime appends a cpuTime under a certain timestamp to record.
+// If the timestamp already exists in tsItems, then cpuTime will be replaced.
```

Comment:
It does not appear in current implementation, because we will **process the CPU data first** and then the StmtStats data.

But for the general purpose of the data model, this is reserved.